rust-libretro/src/environment.rs

```diff
@@ -282,30 +330,21 @@ pub unsafe fn set_hw_render(callback: retro_environment_t, data: retro_hw_render
 #[proc::context(GenericContext)]
 #[proc::context(OptionsChangedContext)]
 #[allow(clippy::needless_lifetimes)]
-pub unsafe fn get_variable<'a>(callback: retro_environment_t, key: &'a str) -> Option<&'a str> {
-    let key = match CString::new(key) {
-        Ok(key) => key,
-        Err(err) => {
-            #[cfg(feature = "log")]
-            log::error!("{}", err);
-
-            return None;
-        }
-    };
+pub unsafe fn get_variable<'a>(
```

Comment:
As far as I can tell RetroArch does not implement this feature. I don’t know whether `libretro.h` is outdated or RetroArch is missing an implementation to return this complete environment string.

Currently I intend to change the code around those environment calls like this:
```rust
// When key is None, sets `retro_variable.key = NULL`, otherwise uses the supplied key
fn get_variable_or_environment<'a>(
    callback: retro_environment_t,
    key: Option<&'a str>,
) -> Result<Option<&'a str>, EnvironmentCallError>

// same as get_variable_or_environment(cb, Some(key))
fn get_variable<'a>(
    callback: retro_environment_t,
    key: &'a str,
) -> Result<Option<&'a str>, EnvironmentCallError>

// calls get_variable_or_environment(cb, None) and parses the value into a key-value map
fn get_all_variables<'a>(
    callback: retro_environment_t,
) -> Result<Option<HashMap<&'a str, &'a str>>, EnvironmentCallError>

// calls RETRO_ENVIRONMENT_SET_VARIABLE with NULL as argument to test for support, as per libretro.h:
// If data is NULL, no changes will be registered and the callback will return true; an implementation may therefore pass NULL in order to test whether the callback is supported.
fn supports_set_variable(callback: retro_environment_t) -> bool
// RetroArch also implements the same behaviour for `RETRO_ENVIRONMENT_GET_VARIABLE`, so there’s also
fn supports_get_variable(callback: retro_environment_t) -> bool
```
That gives `Ok(Some(value))` (frontend returned gave us a `var.value != NULL`), `Ok(None)` (frontend gave us `var.value = NULL`, i.e. variable is unknown) and `Err(_)` in case the callback returns false or some other error happened like failing string conversion.